unstructured/ingest/connector/pinecone.py

```diff
@@ -0,0 +1,127 @@
+import json
+import multiprocessing as mp
+import typing as t
+from dataclasses import dataclass
+
+from unstructured.ingest.error import DestinationConnectionError, WriteError
+from unstructured.ingest.interfaces import (
+    BaseConnectorConfig,
+    BaseDestinationConnector,
+    BaseIngestDoc,
+    BaseSessionHandle,
+    ConfigSessionHandleMixin,
+    IngestDocSessionHandleMixin,
+    WriteConfig,
+)
+from unstructured.ingest.logger import logger
+from unstructured.staging.base import flatten_dict
+from unstructured.utils import requires_dependencies
+
+
+@dataclass
+class PineconeSessionHandle(BaseSessionHandle):
+    service: "pinecone.Index"  # noqa: F821
+
+
+@DestinationConnectionError.wrap
+@requires_dependencies(["pinecone"], extras="pinecone")
+def create_pinecone_object(api_key, index_name, environment):
+    import pinecone
+
+    pinecone.init(api_key=api_key, environment=environment)
+    index = pinecone.Index(index_name)
+    logger.debug(f"Connected to index: {pinecone.describe_index(index_name)}")
+    return index
+
+
+@dataclass
+class SimplePineconeConfig(ConfigSessionHandleMixin, BaseConnectorConfig):
+    api_key: str
+    index_name: str
+    environment: str
+
+    def create_session_handle(self) -> PineconeSessionHandle:
+        service = create_pinecone_object(self.api_key, self.index_name, self.environment)
+        return PineconeSessionHandle(service=service)
+
+
+@dataclass
+class PineconeWriteConfig(IngestDocSessionHandleMixin, WriteConfig):
+    connector_config: SimplePineconeConfig
+    batch_size: int = 50
+    num_processes: int = 1
+
+    @DestinationConnectionError.wrap
+    @requires_dependencies(["pinecone"], extras="pinecone")
```

Comment:
A method responsible for writing the content should live within the destination connector itself rather than on the write config. Ideally this is only responsible for the values and possible some parsing of them.

Reply:
Agreed, addressed with [move upsert method](https://github.com/Unstructured-IO/unstructured/pull/1774/commits/29758f88ae6b539d315208c4da57e7ec23a937ce)

The commit does not seem to break session handles as well: 
- when `number of elements to upload (1404) > num_processes (12, 5, 3)` ,  `num_processes` sessions were created . 
- when `num_processes (5) > number of elements to upload (2)` ,  1 session was created.